doc/source/tutorials/configuration.rst
.. role:: bolditalic
.. role:: hlight 

.. _config:

**********************
The Configuration File
**********************

.. _sample_config:

Configuration files are used to control analysis settings for |dpp-msbayes|_
and |msbayes|_.
|pmb|_ allows you to use |dpp-msbayes|_ and |msbayes|_ configuration files (and
their respective models) interchangeably.
Here is an example of a configuration file for |dpp-msbayes|_::

    concentrationShape = 1000.0
    concentrationScale = 0.00437
    thetaShape = 4.0
    thetaScale = 0.001
    ancestralThetaShape = 0
    ancestralThetaScale = 0
    thetaParameters = 000
    tauShape = 1.0
    tauScale = 0.02
    timeInSubsPerSite = 1
    bottleProportionShapeA = 0
    bottleProportionShapeB = 0
    bottleProportionShared = 0
    migrationShape = 0
    migrationScale = 0
    numTauClasses = 0
    
    BEGIN SAMPLE_TBL
    species-1	locus-1	1.0	1.0	10	8	32.42	389	0.27	0.24	0.26	species-1-locus-1.fasta
    species-1	locus-2	1.0	1.0	8	6	5.51	500	0.25	0.22	0.24	species-1-locus-2.fasta
    species-1	locus-3	1.0	1.0	6	8	8.38	524	0.26	0.23	0.26	species-1-locus-3.fasta
    species-1	locus-4	1.0	1.0	8	10	5.20	345	0.25	0.23	0.24	species-1-locus-4.fasta
    species-1	locus-5	1.0	1.0	8	8	29.59	417	0.27	0.23	0.21	species-1-locus-5.fasta
    species-1	mito-1	0.25	4.0	5	5	8.15	600	0.22	0.24	0.27	species-1-mito-1.fasta
    species-2	locus-1	1.0	1.0	6	10	7.53	400	0.25	0.24	0.26	species-2-locus-1.fasta
    species-2	locus-3	1.0	1.0	10	8	11.14	550	0.27	0.22	0.24	species-2-locus-3.fasta
    species-2	locus-4	1.0	1.0	8	8	9.39	350	0.24	0.24	0.23	species-2-locus-4.fasta
    species-2	locus-5	1.0	1.0	10	10	13.32	450	0.26	0.24	0.22	species-2-locus-5.fasta
    species-2	mito-1	0.25	4.0	4	5	7.59	549	0.23	0.26	0.23	species-2-mito-1.fasta
    species-3	locus-1	1.0	1.0	10	6	17.03	367	0.25	0.23	0.27	species-3-locus-1.fasta
    species-3	locus-3	1.0	1.0	8	10	59.17	541	0.26	0.22	0.25	species-3-locus-3.fasta
    species-3	locus-4	1.0	1.0	6	8	6.90	333	0.28	0.23	0.21	species-3-locus-4.fasta
    species-3	mito-1	0.25	4.0	5	4	11.42	587	0.22	0.22	0.25	species-3-mito-1.fasta
    END SAMPLE_TBL

A configuration file has two parts:

#. The preamble_ containing keyword arguments for specifying prior probability
   distributions of model parameters.
#. The sample_table_ containing information and parameter values for each sequence
   alignment. This section is delimited by the ``BEGIN SAMPLE_TBL`` and ``END
   SAMPLE_TBL`` lines.


.. _preamble:

The Preamble
============

There are two main differences to the preamble between |pmb|_ and
|msbayes|_.

#. The |dpp-msbayes|_ preamble gives you more (and different) options to
   control the number of parameters in the model and the priors on those
   parameters.
#. The |msbayes|_- and |dpp-msbayes|_-style preambles are processed differently
   in |pmb|_: The keywords in the preamble are case insensitive, and if any
   unrecognized keywords are encountered, an error is reported and the process
   exits (i.e., crashes).
   
   In |msbayes|_, the keywords are case-sensitive and, more importantly, the
   the default value for a setting is :hlight:`quietly` used when any options
   are mis-typed.
   For example, if you specify ``uppertheta = 0.01`` in an |msbayes|_ config
   file (note the lower case "t"), it will quietly use the default setting for
   ``upperTheta`` and :hlight:`not` report any warning or error.
   So, if you prefer the model implemented in |msbayes|_, |pmb|_ is useful as a
   "safer" interface for it.

Below we walk through all of the options for the |dpp-msbayes|_ preamble.
If you want to use the |msbayes|_ model with |pmb|_, please see the
`msBayes documentation
<https://docs.google.com/document/d/15heQlz60cGe6GWKcXqf1AIYMBZ6p2sKuGct-EoEHiNU/edit>`_
for information about the preamble.


.. contents:: Preamble Elements
    :local:


.. _concentration_parameter_prior:

``concentrationShape`` / ``concentrationScale``
-----------------------------------------------
    
The shape and scale parameters of a gamma-distributed prior on the
concentration parameter of the Dirichlet-process prior on divergence models.

See the ":ref:`dpp`" section for more information on the Dirichlet process
and the concentration parameter that controls it.


.. _theta_parameterization:

``thetaParameters``
-------------------

|dpp-msbayes|_ gives you full control over the parameterization of the
population sizes for each pair of populations.
We will refer to the sizes of the ancestral and two descendant populations as
:math:`\ancestralTheta{}`, :math:`\descendantTheta{1}{}`, and
:math:`\descendantTheta{2}{}`.

This setting is controlled by a sequence of three integers that always starts
with ``0`` and increments by 1 whenever a free parameter is added.
The first, second, and third integers represent :math:`\descendantTheta{1}{}`,
:math:`\descendantTheta{2}{}`, and :math:`\ancestralTheta{}`, respectively.

``000`` is one extreme, where :math:`\ancestralTheta{}`,
:math:`\descendantTheta{1}{}`, and :math:`\descendantTheta{2}{}` are all
constrained to be equal for each population pair (population sizes will still
vary among pairs).

``012`` is the other extreme, where :math:`\ancestralTheta{}`,
:math:`\descendantTheta{1}{}`, and :math:`\descendantTheta{2}{}` are all
estimated as independent parameters for each pair of populations.
This is most similar to the original |msbayes|_.
However, the descendant population sizes are *not* constrained to be negatively
correlated, as they are in |msbayes|_ (see :cite:`Oaks2014dpp` and
:cite:`Oaks2012`).

Another example is ``001``: the descendant populations share the same size
parameter, but the ancestral population size is free to vary.

For ``011`` and ``010``, one of the descendant population is constrained to the
same size as the ancestral, and the other is free to vary.


.. _theta_prior:

``thetaShape`` / ``thetaScale``
-------------------------------

These settings define the shape and scale parameters of a gamma prior on the
effective population sizes. Population sizes are scaled by the per-site
mutation rate (:math:`\mu`): :math:`4\effectivePopSize\mutationRate`.


.. _ancestral_theta_prior:

``ancestralThetaShape`` / ``ancestralThetaScale``
-------------------------------------------------

If these settings are both provided, and are both positive, they define the
shape and scale parameters of a gamma prior on the effective population size of
the ancestral population of each pair of populations.

If they are excluded, or both are zero, the :ref:`theta_prior`
settings are used for the gamma prior on ancestral population sizes (i.e.,
all population size parameters have the same prior).

.. note::
    :class: keypoint

    If you specify a setting such as ``000`` for the
    :ref:`theta_parameterization` option, the ancestral population size is
    constrained to be the same as the descendant populations, and thus the
    :ref:`ancestral_theta_prior` options will be overridden and have no affect.


.. _divergence_time_prior:

``tauShape`` / ``tauScale``
---------------------------

These settings define the shape and scale parameters of a gamma prior
on divergence times. 
See the :ref:`timescale_setting` setting for the information on
the time units.


.. _timescale_setting:

``timeInSubsPerSite``
---------------------

This setting controls the time scale of the model and has two settings:

*   ``timeInSubsPerSite = 1``: Time units are in expected substitutions per
    site.
    For example, a divergence of 0.05 means that, on average, 5% of sites have
    changed since the populations diverged (so you expect 10% divergence
    between the populations since the population divergence).
    Thus, you can convert these units to the number of generations or years by
    dividing by the mutation rate.

*   ``timeInSubsPerSite = 0``: Time units are in coalescent units,
    :math:`\globalcoalunit` generations, where :math:`\globalpopsize` is the
    size of a constant reference population based on the mean of the prior on
    theta (defined by the :ref:`theta_prior` settings ).

    If we use :math:`\globaltheta` to represent the mean of the theta prior,
    then
    
    .. math::
        \globalcoalunit = \frac{\globaltheta}{\mutationRate},

    where :math:`\mutationRate` is the per-site mutation rate.  Thus, you can
    convert these ":math:`\globalcoalunit` generations" units to the number of
    generations by assuming a mutation rate and multiplying by
    :math:`(\globaltheta/\mutationRate)`.  See :cite:`Oaks2014dpp` for more
    details.

    |pmb|_ supports the use of the ``timeInSubsPerSite`` option for |msbayes|_
    configuration files as well.


    .. note::
        :hlight:`Why use the mean of the prior on theta to scale time?`
        I have no idea.
        This is legacy from |msbayes|_, and is the default setting.
        However, I strongly discourage using this time scale, because it makes it
        very difficult to compare results across analyses with different settings
        for the theta_prior_.
        It also requires you to re-scale the divergence_time_prior_ every time
        you change the theta_prior_.
        Having time in units of expected substitutions per site is much more
        straight forward (i.e., ``timeInSubsPerSite = 1``).


.. _bottleneck_prior:

``bottleProportionShapeA`` / ``bottleProportionShapeB``
-------------------------------------------------------

If both are positive, these settings define a beta prior on the magnitude of a
post-divergence bottleneck in each of the descendant populations.
``bottleProportionShapeA`` and ``bottleProportionShapeB`` correspond to the
shape parameters alpha and beta, respectively, of the beta prior.

The bottleneck magnitude is the proportion of the effective population size
that remains during the bottleneck.
For example, a value of 0.95 would mean that bottleneck reduces the effective
population size by 5%.

If either or both are zero or less, there is no post-divergence population
bottleneck in the descendant populations (i.e., the bottleneck-magnitude
parameters, along with the timing of each bottleneck, are removed from the
model).

.. note::
    There are also parameters in the model for the timing of the end of the
    bottleneck (it begins at speciation in forward time). There is one of these
    parameters for each pair of populations (i.e., the descendant populations
    of each pair share the same bottleneck-end-time parameter).
    Thus if either or both of the :ref:`bottleneck_prior` settings are zero or
    less, you are also removing these bottleneck timing parameters from the
    model.
    This means you are removing :math:`3\npairs{}` parameters from the model,
    where ":math:`\npairs{}`" is the number of pairs of populations.


.. _bottleneck_parameterization:

``bottleProportionShared``
--------------------------

If ``bottleProportionShared = 0``, then there are two free bottleneck-magnitude
parameters for each population pair (one for each descendant population).
If ``bottleProportionShared = 1``, then there is one bottleneck-magnitude
parameter for each population pair (i.e., the descendant populations of each
pair share the same bottleneck magnitude; the bottleneck magnitude still varies
among the pairs).

.. note::
    This setting is overridden if either or both of the
    ``bottleProportionShapeA`` and ``bottleProportionShapeB`` settings is zero
    or less (because then there is no bottleneck at all).


``migrationShape`` / ``migrationScale``
---------------------------------------

These settings define the shape and scale parameters of a gamma prior on the
symmetric migration between the descendant populations of each pair (in units
of the number of gene copies per generation).

If either or both settings are zero or less, there is no migration in the
model.


``recombinationShape`` / ``recombinationScale``

CHANGES.txt
Version 0.3.0
=============

Changes
-------
  - Updated the version of ``dpp-msprior`` binary (for both mac and linux) that
    is bundled with the package. This version of ``dpp-msprior`` is from
    version 0.2 of the ``dpp-msbayes`` package and was updated to create a
    lower limit (0.000000000001) for values of theta. In rare cases, a theta
    value of zero would cause the coalescent simulator ``msDQH`` to crash.
    Any previous analyses that did not crash should not be affected by
    this change.
  - The newer version of ``dpp-msprior`` also changes the weirdness from the
    original ``msBayes`` where there was a check for small (i.e., 0.0001)
    divergence times. In such simulations, the div time was set to this
    arbitrary lower bound, and the bottleneck time was set to 0.5 of this.  I
    am guessing this was to prevent unrealistic (and numerically unstable?)
    changes in pop size. However, 0.0001 can be thousands of generations which
    is not trivial. Also, rather than this weird hack of the bottleneck time,
    it seems much better to simply have no bottleneck if the div time is
    essentially zero. Accordingly, I lowered the threshold and simply "turn
    off" the bottleneck if the time is below it (I no longer adjust the div
    time or bottleneck time). This change should have very little affect on
    most analyses.

Bug Fixes
---------
  - As mentioned in the change above, in rare cases, theta values of zero drawn
    from the prior via ``dpp-msprior`` would cause ``msDQH`` to crash. The new
    version of the ``dpp-msprior`` binaries should prevent this. Any previous
    analyses that did not crash were not affected by this bug.


Version 0.2.2
=============

Changes
-------
  - End-user CLI scripts log additional information.
  - ``dmc.py`` will now set the number of standardizing samples equal to the
    number of prior samples, if the former is larger than the latter.

New Features
------------
  - New management of paths to dpp-msbayes and msBayes executables. Previously,
    the package only used the executables bundled with the package, which
    required the package to be installed via ``python setup.py develop``. Now,
    the package will also look for executables on the system.
  - Updating ``setup.py`` so that the bundled executables will be installed
    along with end-user scripts via ``python setup.py install`` after checking
    that all the executables work on the system.
  - 32-bit Linux versions of the dpp-msbayes/msBayes executables are now
    bundled with the package.

Bug Fixes
---------
  - Fixed a bug where ``MsBayesWorker`` instances with a sample size of zero
    would hang indefinitely. A ValueError is now raised if an ``MsBayesWorder``
    is initiated with an integer less than one for the sample size.


Version 0.2.1
=============

Changes
-------
  - Adding working version of documentation.
  - Overhauled package-wide and CLI-script logging management.

Bug Fixes
---------
  - Fixing bug in ``dmc.py`` script, where the wrong value for the number of
    standardizing samples was being written to the run summary file.


Version 0.2.0
=============

New Features
------------
  - New options added to package and scripts for the sort index. These allow
    summary statistics to be grouped by taxon, but without re-sorting.

Changes
-------
  - The new default option for the sort index for the entire package and for
    end-user scripts is to not group or re-sort the summary statistics of each
    alignment. The re-sorting that was done prior to this was NOT VALID,
    because the summary statics calculated from the different alignments are
    NOT exchangeable.


pymsbayes/config.py
#! /usr/bin/env python

import os
import sys
import re
import math
from cStringIO import StringIO
import ConfigParser

from pymsbayes import fileio
from pymsbayes.utils import probability, errors
from pymsbayes.utils.containers import OrderedDict
from pymsbayes.utils.messaging import get_logger

_LOG = get_logger(__name__)

class MsBayesConfig(object):
    _table_begin_pattern = re.compile(r'^begin\s*sample_tbl$', re.IGNORECASE)
    _table_end_pattern = re.compile(r'^end\s*sample_tbl$', re.IGNORECASE)

    def __init__(self, cfg_file):
        self.theta = None
        self.d_theta = None
        self.a_theta = None
        self.tau = None
        self.time_in_subs_per_site = None
        self.migration = None
        self.recombination = None
        self.psi = None
        self.dpp_concentration = None
        self.npairs = None
        self.implementation = 'old'
        self.div_model_prior = None
        self.theta_parameters = None
        self.bottle_proportion = None
        self.bottle_proportion_shared = None
        self.taxa = []
        self.sample_table = None
        self._parse_config(cfg_file)

    @classmethod
    def is_config(cls, cfg_file):
        cfg_stream, close = fileio.process_file_arg(cfg_file)
        for i in range(100):
            try:
                line = cfg_stream.next()
                if cls._table_begin_pattern.match(line.strip()):
                    if close:
                        cfg_stream.close()
                    return True
            except:
                if close:
                    cfg_stream.close()
                return False
        if close:
            cfg_stream.close()
        return False

    def _parse_config(self, cfg):
        preamble, table = self._split_config(cfg)
        preamble.seek(0)
        table.seek(0)
        self._parse_table(table)
        self._set_priors(preamble)

    def _split_config(self, cfg):
        cfg_stream, close = fileio.process_file_arg(cfg)
        preamble = StringIO()
        table = StringIO()
        preamble.write('[preamble]\n')
        table_started = False
        for i, line in enumerate(cfg_stream):
            if self._table_end_pattern.match(line.strip()):
                table.write(line)
                break
            if self._table_begin_pattern.match(line.strip()):
                table_started = True
                table.write(line)
                continue
            if table_started:
                table.write(line)
            else:
                preamble.write(line)
        if close:
            cfg_stream.close()
        return preamble, table

    def _parse_table(self, table):
        self.sample_table = SampleTable(table)
        self.taxa = list(self.sample_table.taxa)
        self.npairs = len(self.taxa) 

    def _get_gamma_or_uniform_distribution(self, shape, scale):
        if shape > 0.0 and scale > 0.0:
            return probability.GammaDistribution(shape, scale)
        else:
            a = math.fabs(shape)
            b = math.fabs(scale)
            if a < b:
                return probability.ContinuousUniformDistribution(a, b)
            else:
                return probability.ContinuousUniformDistribution(b, a)

    def _get_gamma_or_uniform_settings(self, distribution):
        if isinstance(distribution, probability.GammaDistribution):
            return distribution.shape, distribution.scale
        return -distribution.minimum, -distribution.maximum

    def _set_priors(self, preamble):
        kwargs = self._parse_preamble(preamble)
        psi = int(kwargs.get('numtauclasses', 0))
        time_units = int(kwargs.get('timeinsubspersite', 0))
        self.time_in_subs_per_site = bool(time_units)
        for k in ['concentrationshape', 'concentrationscale', 'taushape',
                'tauscale', 'thetashape', 'thetascale', 'ancestralthetashape',
                'ancestralthetascale', 'thetaparameters']:
            if kwargs.has_key(k):
                self.implementation = 'new'
        
        if self.implementation == 'new':
            dpp_concentration_shape = float(kwargs.get('concentrationshape',
                    0.0))
            dpp_concentration_scale = float(kwargs.get('concentrationscale',
                    0.0))
            if psi != 0:
                self.psi = probability.DiscreteUniformDistribution(psi, psi)
                self.dpp_concentration = None
                self.div_model_prior = 'constrained'
            elif (dpp_concentration_shape > 0.0) and (
                    dpp_concentration_scale > 0.0):
                self.div_model_prior = 'dpp'
                self.dpp_concentration = probability.GammaDistribution(
                        dpp_concentration_shape,
                        dpp_concentration_scale)
            elif (dpp_concentration_shape > -1.0) and (
                    dpp_concentration_scale > -1.0):
                self.div_model_prior = 'uniform'
                self.dpp_concentration = None
            else:
                self.div_model_prior = 'psi'
                self.dpp_concentration = None
                self.psi = probability.DiscreteUniformDistribution(1,
                        self.npairs)

            tau_shape = float(kwargs.get('taushape', 1.0))
            tau_scale = float(kwargs.get('tauscale', 2.0))
            self.tau = self._get_gamma_or_uniform_distribution(tau_shape,
                    tau_scale)
            theta_shape = float(kwargs.get('thetashape', 1.0))
            theta_scale = float(kwargs.get('thetascale', 0.001))
            anc_theta_shape = float(kwargs.get('ancestralthetashape', 0.0))
            anc_theta_scale = float(kwargs.get('ancestralthetascale', 0.0))
            theta_params = kwargs.get('thetaparameters', '012')
            self.theta_parameters = [int(i) for i in list(theta_params)]
            self.theta = None
            self.d_theta = self._get_gamma_or_uniform_distribution(theta_shape,
                    theta_scale)
            if ((anc_theta_shape > 0.0) and (anc_theta_scale > 0.0)) and (
                    (theta_params == '001') or (theta_params == '012')):
                self.a_theta = self._get_gamma_or_uniform_distribution(
                        anc_theta_shape,
                        anc_theta_scale)
            else:
                self.a_theta = self._get_gamma_or_uniform_distribution(
                        theta_shape,
                        theta_scale)
            mig_shape = float(kwargs.get('migrationshape', 0.0))
            mig_scale = float(kwargs.get('migrationscale', 0.0))
            if (mig_shape > 0.0) and (mig_scale > 0.0):
                self.migration = probability.GammaDistribution(mig_shape,
                        mig_scale)
            else:
                self.migration = probability.ContinuousUniformDistribution(0.0,
                        0.0)
            rec_shape = float(kwargs.get('recombinationshape', 0.0))
            rec_scale = float(kwargs.get('recombinationscale', 0.0))
            if (rec_shape > 0.0) and (rec_scale > 0.0):
                self.recombination = probability.GammaDistribution(rec_shape,
                        rec_scale)
            else:
                self.recombination = probability.ContinuousUniformDistribution(
                        0.0, 0.0)
            bottle_a = float(kwargs.get('bottleProportionShapeA', 0.0))
            bottle_b = float(kwargs.get('bottleProportionShapeB', 0.0))
            if (bottle_a > 0.0) and (bottle_b > 0.0):
                self.bottle_proportion = probability.BetaDistribution(bottle_a,
                        bottle_b)
            else:
                self.bottle_proportion = None
            bottle_shared = int(kwargs.get('bottleProportionShared', 0))
            self.bottle_proportion_shared = bool(bottle_shared)
            return

        ltau = float(kwargs.get('lowertau', 0.0))
        utau = float(kwargs.get('uppertau', 0.0))
        psi = int(kwargs['numtauclasses'])
        ltheta = float(kwargs.get('lowertheta', 0.0))
        utheta = float(kwargs['uppertheta'])
        anc_scalar = float(kwargs['upperancpopsize'])
        lmig = float(kwargs.get('lowermig', 0.0))
        umig = float(kwargs.get('uppermig', 0.0))
        lrec = float(kwargs.get('lowerrec', 0.0))
        urec = float(kwargs.get('upperrec', 0.0))
        if psi == 0:
            self.psi = probability.DiscreteUniformDistribution(1, self.npairs)
            self.div_model_prior = 'psi'
        else:
            self.psi = probability.DiscreteUniformDistribution(psi, psi)
            self.div_model_prior = 'constrained'
        self.tau = probability.ContinuousUniformDistribution(ltau, utau)
        self.theta = probability.ContinuousUniformDistribution(ltheta, utheta)
        self.a_theta = probability.ContinuousUniformDistribution(ltheta,
                utheta*anc_scalar)
        self.d_theta = probability.BetaDistribution(1, 1, 2)
        self.migration = probability.ContinuousUniformDistribution(lmig, umig)
        self.recombination = probability.ContinuousUniformDistribution(lrec,
                urec)

    def _parse_preamble(self, preamble):
        cfg = ConfigParser.RawConfigParser()
        cfg.readfp(preamble)
        d = {}
        for tup in cfg.items('preamble'):
            d[tup[0].lower()] = tup[1]
        return d

    def _get_old_settings(self):
        d = {}
        d['upperTheta'] = self.theta.maximum
        d['lowerTheta'] = self.theta.minimum
        d['upperTau'] = self.tau.maximum
        d['lowerTau'] = self.tau.minimum
        d['timeInSubsPerSite'] = int(self.time_in_subs_per_site)
        d['upperAncPopSize'] = self.a_theta.maximum /self.theta.maximum
        if self.div_model_prior.lower() == 'constrained':
            assert self.psi.maximum == self.psi.minimum
            d['numTauClasses'] = self.psi.maximum
        else:
            d['numTauClasses'] = 0
        d['upperMig'] = self.migration.maximum
        d['upperRec'] = self.recombination.maximum
        d['constrain'] = 0
        d['subParamConstrain'] = '111111111'
        return d
    
    def _get_new_settings(self):
        d = {}
        if self.dpp_concentration:
            d['concentrationShape'] = self.dpp_concentration.shape
            d['concentrationScale'] = self.dpp_concentration.scale
        elif self.div_model_prior.lower() == 'psi':
            d['concentrationShape'] = -1.0
            d['concentrationScale'] = -1.0
        else:
            d['concentrationShape'] = 0.0
            d['concentrationScale'] = 0.0
        d['thetaShape'], d['thetaScale'] = \
                self._get_gamma_or_uniform_settings(self.d_theta)
        d['ancestralThetaShape'], d['ancestralThetaScale'] = \
                self._get_gamma_or_uniform_settings(self.a_theta)
        d['thetaParameters'] = ''.join([str(x) for x in self.theta_parameters])
        d['tauShape'], d['tauScale'] = \
                self._get_gamma_or_uniform_settings(self.tau)
        d['timeInSubsPerSite'] = int(self.time_in_subs_per_site)
        if self.bottle_proportion:
            d['bottleProportionShapeA'] = self.bottle_proportion.alpha
            d['bottleProportionShapeB'] = self.bottle_proportion.beta
        else:
            d['bottleProportionShapeA'] = 0.0
            d['bottleProportionShapeB'] = 0.0
        if self.bottle_proportion_shared:
            d['bottleProportionShared'] = 1
        else:
            d['bottleProportionShared'] = 0
        if self.div_model_prior.lower() == 'constrained':
            assert self.psi.maximum == self.psi.minimum
            d['numTauClasses'] = self.psi.maximum
        else:
            d['numTauClasses'] = 0
        d['migrationShape'], d['migrationScale'] = \
                self._get_gamma_or_uniform_settings(self.migration)
        d['recombinationShape'], d['recombinationScale'] = \
                self._get_gamma_or_uniform_settings(self.recombination)
        d['constrain'] = 0
        d['subParamConstrain'] = '111111111'
        return d

    def get_settings(self):
        if self.implementation.lower() == 'old':
            return self._get_old_settings()
        return self._get_new_settings()

    def get_preamble(self):
        s = StringIO()
        settings = self.get_settings()
        for k in sorted(settings.iterkeys()):
            s.write('{0} = {1}\n'.format(k, settings[k]))
        return s.getvalue()

    def equal_sample_table(self, config):
        if not isinstance(config, MsBayesConfig):
            config = MsBayesConfig(config)
        return self.sample_table.equals(config.sample_table)

    def __str__(self):
        return '\n'.join([self.get_preamble() + str(self.sample_table)])

    def write(self, file_obj):
        out, close = fileio.process_file_arg(file_obj, 'w')
        out.write(self.__str__())
        if close:
            out.close()


class SampleTable(object):
    _begin_pattern = re.compile(r'^begin\s*sample_tbl$', re.IGNORECASE)
    _end_pattern = re.compile(r'^end\s*sample_tbl$', re.IGNORECASE)

    def __init__(self, config_file):
        self.alignments = None
        self._ordering = []
        self._parse_table(config_file)

    def _parse_table(self, config_file):
        self.alignments = OrderedDict()
        cfg_stream, close = fileio.process_file_arg(config_file)
        try:
            table_started = False
            table_finished = False
            row_num = 0
            for i, l in enumerate(cfg_stream):
                line = l.strip()
                if self._end_pattern.match(line):
                    if not table_started:
                        raise errors.SampleTableError(
                                'hit end of sample table before beginning')
                    if len(self.alignments) < 1:
                        raise errors.SampleTableError(
                                'no rows found in sample table')
                    table_finished = True
                    break
                if self._begin_pattern.match(line):
                    table_started = True
                    continue
                if not table_started:
                    continue
                if (line == '') or (line.startswith('#')):
                    continue
                row_num += 1
                try:
                    al = AlignmentConfig(line)
                except errors.SampleTableRowError as e:
                    _LOG.error('sample table row {0} is invalid'.format(
                            row_num))
                    raise e
                if not al.taxon_name in self.alignments:
                    self.alignments[al.taxon_name] = OrderedDict()
                    self.alignments[al.taxon_name][al.locus_name] = al
                    self._ordering.append((al.taxon_name, al.locus_name))
                    continue
                if al.locus_name in self.alignments[al.taxon_name]:
                    raise errors.SampleTableError('locus {0} found twice '
                            'for taxon {1} at row {2} of sample '
                            'table'.format(al.locus_name, al.taxon_name,
                                    row_num))
                self.alignments[al.taxon_name][al.locus_name] = al
                self._ordering.append((al.taxon_name, al.locus_name))
            if not table_started:
                raise errors.SampleTableError('no sample table found')
            if not table_finished:
                raise errors.SampleTableError('no end of table found')
        finally:
            if close:
                cfg_stream.close()

    def _get_taxa(self):
        return self.alignments.keys()

    taxa = property(_get_taxa)

    def _get_loci(self):
        l = []
        for t, d in self.alignments.iteritems():
            for locus in d.iterkeys():
                if not locus in l:
                    l.append(locus)
        return l

    loci = property(_get_loci)

    def _get_number_of_taxa(self):
        return len(self.taxa)

    npairs = property(_get_number_of_taxa)

    def get_sample_table_string(self):    
        return '\n'.join(('BEGIN SAMPLE_TBL',
                '\n'.join((str(self.alignments[t][l]) for t, l in self._ordering)),
                'END SAMPLE_TBL'))

    def __str__(self):
        return self.get_sample_table_string()

    def equals(self, other):
        if not isinstance(other, SampleTable):
            return False
        if len(self.alignments) != len(other.alignments):
            return False
        for i1, i2 in zip(self.alignments.items(), other.alignments.items()):
            if i1[0] != i2[0]:
                return False
            if len(i1[1]) != len(i2[1]):
                return False
            for t1, t2 in zip(i1[1].items(), i2[1].items()):
                if t1[0] != t2[0]:
                    return False
                if not t1[1].equals(t2[1]):
                    return False
        if self._ordering != self._ordering:
            return False
        return True
                                   

class AlignmentConfig(object):     

    def __init__(self, sample_table_row = None):
        self.taxon_name = None
        self.locus_name = None
        self._ploidy_multiplier = None
        self._mutation_rate_multiplier = None
        self._number_of_gene_copies = None
        self._kappa = None
        self._length = None
        self._base_frequencies = None
        self.path = None
        if sample_table_row != None:
            self._parse_sample_table_row(sample_table_row)

    def _parse_sample_table_row(self, sample_table_row):
        if isinstance(sample_table_row, str):
            row = [e.strip() for e in sample_table_row.split()]
        else:
            row = list(sample_table_row)
        if len(row) != 12:
            raise errors.SampleTableRowError(
                    'sample table row has {0} columns'.format(len(row)))
        self.taxon_name, self.locus_name = row[0:2]
        self._ploidy_multiplier, self._mutation_rate_multiplier = row[2:4]
        self._number_of_gene_copies = tuple(row[4:6])
        self._kappa, self._length = row[6:8] 
        self._base_frequencies = tuple(row[8:11])
        self.path = row[11]

    def _get_ploidy_multiplier(self):
        return float(self._ploidy_multiplier)

    ploidy_multiplier = property(_get_ploidy_multiplier)

    def _get_mutation_rate_multiplier(self):
        return float(self._mutation_rate_multiplier)

    mutation_rate_multiplier = property(_get_mutation_rate_multiplier)

    def _get_number_of_gene_copies(self):
        return tuple(int(x) for x in self._number_of_gene_copies)

    number_of_gene_copies = property(_get_number_of_gene_copies)

    def _get_kappa(self):
        return float(self._kappa)

    kappa = property(_get_kappa)

    def _get_length(self):
        return int(self._length)

    length = property(_get_length)

    def _get_base_frequencies(self):
        return tuple(float(x) for x in self._base_frequencies)

    base_frequencies = property(_get_base_frequencies)

    def get_sample_table_row_elements(self):
        return(self.taxon_name,
                self.locus_name,
                self.ploidy_multiplier,
                self.mutation_rate_multiplier,
                self.number_of_gene_copies[0],
                self.number_of_gene_copies[1],
                self.kappa,
                self.length,
                self.base_frequencies[0],
                self.base_frequencies[1],
                self.base_frequencies[2],
                self.path)

    def get_sample_table_row_element_strings(self):
        return(self.taxon_name,
                self.locus_name,
                self._ploidy_multiplier,
                self._mutation_rate_multiplier,
                self._number_of_gene_copies[0],
                self._number_of_gene_copies[1],
                self._kappa,
                self._length,
                self._base_frequencies[0],
                self._base_frequencies[1],
                self._base_frequencies[2],
                self.path)

    def get_sample_table_row_string(self):
        return '\t'.join([str(e) for e in self.get_sample_table_row_element_strings()])

    def equals(self, other):
        return self.__dict__ == other.__dict__
        
    def __str__(self):
        return self.get_sample_table_row_string()

